### PR TITLE
[Perf] Give a prefill-only batch the model state's number of logit rows

### PR DESCRIPTION
**Pull request by mmastrac**
## Purpose

Performance win: a prefill-only batch in a diffusion model (no draft tokens scheduled) hands every request one logit row regardless of what the model state says it samples per step.

DiffusionGemma's prefill step never reads them so every prefill has to take an unnecessary full pass over the 262144-row vocabulary matrix, about an extra 8.8 ms on a GB10.

Split out of #57250.

## Test plan

Hardware: 1 DGX Spark.

Model: `nvidia/diffusiongemma-26B-A4B-it-NVFP4`

Serve with: `--diffusion-config '{"canvas_length": 32}' --max-logprobs 32 --enable-prefix-caching --async-scheduling`, before/after patch (or torch profiler)

## Test Results

Before/after with the model runner swapped and nothing else changed:

median read | time
-|-
before, one row per request	| 116.5 ms 
 after	| 106.9 ms 

Before: `compute_logits` runs on every DiffusionGemma prefill step, ~8.8 ms per prefill in the torch profile.

After: no logits pass on prefill steps, decode steps unchanged, 32-way concurrency sweep with logprobs passes. Autoregressive models unchanged and take the same branch with one row per request, as before.

---

<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by hclsys** [comment]
The `num_bonus_tokens = 0` case reads correctly to me, and 8.8 ms of vocab pass per prefill on a
GB10 is worth reclaiming (I have the same card).

One thing I'd check before this lands, on the `not draft_tokens` fast path:

```python
total_num_logits = num_reqs * num_bonus_tokens
expanded_idx_mapping = idx_mapping[:total_num_logits]
expanded_local_pos = torch.zeros(total_num_logits, ...)
```

`idx_mapping` is built with `count=num_reqs` (`model_runner.py:1235`), so it has exactly `num_reqs`
rows. For `num_bonus_tokens > 1` the slice asks for `num_reqs * k` rows and **silently returns
`num_reqs`** — Python slicing doesn't bounds-check — while `expanded_local_pos` and `cu_num_logits`
are both sized `num_reqs * k`. Two arrays that are supposed to line up no longer do, with no error.

The all-zeros `expanded_local_pos` has the same shape of problem. `_expand_idx_mapping_kernel`
(`input_batch.py:687`) fills local positions with `block`, i.e. `0..n-1` per request, and *repeats*
the request index rather than truncating:

```python
tl.store(expanded_idx_mapping_ptr + start_idx + block, req_state_idx, mask=mask)
tl.store(expanded_local_pos_ptr  + start_idx + block, block,         mask=mask)
```

For `k in (0, 1)` the hand-rolled versions and the kernel agree exactly, which is why the diffusion
case you measured is fine. For `k > 1` they don't.

I can't show you a config that reaches this path with `k > 1` — that's the part I'd want your read
on, since you know the diffusion/MTP side better than I do. What makes me raise it rather than
assume it's unreachable is that the `else` branch twelve lines down treats `k > 1` as ordinary
(`total_num_logits = num_reqs * num_bonus_tokens + total_num_draft_tokens`), so the surrounding code
clearly contemplates it.

If it is reachable, calling `expand_idx_mapping(...)` in the fast path too would keep both branches
on the same code. If it isn't, an `assert num_bonus_tokens <= 1` here would be cheap and would stop
the truncation being silent later.

**Comment by mmastrac** [comment]
@hclsys Thanks for the review! Would it be possible to add the `ready` flag to the PR? I'll dive deep on your feedback in the meantime to confirm.

**Comment by hclsys** [comment]
Afraid I can't — I'm an outside contributor here, read-only on the repo, so labels aren't something I can set. Your PR came up in a GB10 search and I reviewed it as another owner of the part, not as anyone with a bit to flip.

For what it's worth, `docs/contributing/README.md` says the `ready` label comes from a reviewer with write access, and `new_pr_bot.yml` gives enabling auto-merge as the other route — `add_label_automerge.yml` adds `ready` itself on `auto_merge_enabled`.

If it helps once you've had a look, I can run it on my Spark. Same part you measured the 8.8 ms on.

**Comment by hclsys** [comment]
Chased the `k > 1` question I left open, so you don't have to spend the time on it.

On `23e26e0588`, the only in-tree override of `num_new_sampled_tokens_per_step` is `diffusion_gemma.py:934`, and it sets **0**. The base default at `model_states/interface.py:237` is **1**. The other hit, `cudagraph_utils.py:276`, only recovers the number arithmetically for graph shapes — it doesn't set the attribute.

So nothing reaches your fast path with `k > 1` today, and 0 and 1 are exactly the two values where the slice and `_expand_idx_mapping_kernel` agree. The truncation is latent, not live.

That changes which of my two suggestions I'd push for. `assert num_bonus_tokens <= 1` is the cheap one: it costs nothing now, documents the assumption the fast path is built on, and turns a silent wrong-shape into a failure on the day someone sets it to 2. Wiring `expand_idx_mapping` into the fast path would buy correctness for a case that doesn't exist yet, at the cost of the launch you're removing — which is the whole point of the PR.

File: vllm/v1/worker/gpu/model_runner.py [hunk 1]
```diff
@@ -1291,16 +1291,20 @@ def prepare_inputs(
...
+            cu_num_logits = (
+                torch.arange(num_reqs + 1, device=self.device, dtype=torch.int32)
+                * num_bonus_tokens
             )
-            expanded_idx_mapping = idx_mapping
+            expanded_idx_mapping = idx_mapping[:total_num_logits]
             expanded_local_pos = torch.zeros(
-                num_reqs, dtype=torch.int32, device=self.device
+                total_num_logits, dtype=torch.int32, device=self.device
             )
```
Comment (LucasWilkinson) [inline vllm/v1/worker/gpu/model_runner.py]: Smaller alternative that keeps the fast path byte-for-byte as on `main` and only takes it when the model state samples one token per step. A diffusion prefill-only batch then falls through to the general branch below, which already gives prefill requests zero logit rows in mixed batches, so no new code path is added.

```suggestion
        if not draft_tokens and self.model_state.num_new_sampled_tokens_per_step == 1:
            # No draft token scheduled (common case).
            total_num_draft_tokens = 0
            total_num_logits = num_reqs
            cu_num_logits_np = np.arange(num_reqs + 1, dtype=np.int32)
            cu_num_logits = torch.arange(
                num_reqs + 1, device=self.device, dtype=torch.int32
            )
            expanded_idx_mapping = idx_mapping
            expanded_local_pos = torch.zeros(
                num_reqs, dtype=torch.int32, device=self.device
            )
```

This also needs the `if draft_tokens:` guard in front of `expand_idx_mapping` (L1360, outside the diff so I can't attach a suggestion there) changed to `if num_draft_tokens_per_req is not None:`, otherwise `expanded_idx_mapping` is unassigned on the fall-through path.

Comment (hclsys) [reply vllm/v1/worker/gpu/model_runner.py]: That guard is nicer than either option I floated — it handles k != 1 instead of rejecting it.

One more thing to change alongside the L1360 guard you flagged, and it moves the opposite way. Today `num_draft_tokens_per_req is not None` is *equivalent* to "`draft_tokens` is non-empty", because only the else branch assigns it. Your change makes that branch reachable with `draft_tokens == {}`, so the two stop being interchangeable:

- L1360 `if draft_tokens:` needs **loosening** to `is not None`, as you said.
- L1324 `adaptive_verification = self.adaptive_verification if num_draft_tokens_per_req is not None else None` needs **tightening**. Otherwise a k != 1 prefill-only batch now enters `compact_batch()` and then `reallocate_drafts()` with all-zero draft counts, where today it skips both — and `reallocate_drafts` overwrites `cu_num_logits` and `query_start_loc` and recomputes `total_num_logits`.

`:1242` already writes the pairing that wants: `if self.adaptive_verification is not None and draft_tokens:`.

What I can't show you is a reachable case. It needs `speculator.enable_adaptive_verification` together with a k != 1 model state, and the only k != 1 in the tree is `diffusion_gemma.py:934`, which is 0. If those two can't coexist then this is latent. But the two guards currently read as if they're the same test, and after this they aren't.

**Review by LucasWilkinson — APPROVED** [review]
LGTM thanks!

File: vllm/v1/worker/gpu/model_runner.py (hunk 1 above)
Comment (hclsys) [reply vllm/v1/worker/gpu/model_runner.py]: Thanks for taking the `draft_tokens` tightening. One more thing on the same fall-through, and this one I could actually run.

The new `if num_draft_tokens_per_req is not None:` also makes `expand_idx_mapping` reachable for the k != 1 no-draft batch, and it is called with `max_expand_len=self.decode_query_len`. That is `self.num_speculative_steps + self.model_state.num_new_sampled_tokens_per_step` (`model_runner.py:430`), so for a diffusion model served without spec decode it is `0 + 0`.

`triton.next_power_of_2(0)` returns **0**, not 1, so `BLOCK_SIZE=0` reaches `tl.arange(0, BLOCK_SIZE)`. On a GB10 (sm_121, torch 2.12.1+cu130, triton from the same build):

```
next_power_of_2(0) = 0
expand_idx_mapping(idx_mapping, total_num_logits=0, cu_num_logits=zeros(5), max_expand_len=0)
  -> triton.compiler.errors.CompilationError
     block = tl.arange(0, BLOCK_SIZE)
     arange's end argument must be greater than the start argument

control, max_expand_len=1:  OK  idx=[0,1,2,3]  local_pos=[0,0,0,0]
```

So either a `decode_query_len` of 0 can't actually occur and I'm missing what keeps it positive for diffusion, or this path raises at compile time for the case the PR is routing here. Worth noting `:791` does `num_reqs = num_tokens // self.decode_query_len` under `uniform_decode`, which also assumes it is non-zero.

If 0 is reachable, guarding the call on `total_num_logits > 0` would sidestep both, since the zero-row batch has nothing to expand anyway.

I can't run a real diffusion checkpoint here, so I haven't seen this fire end to end — only the kernel call with the shapes that path would produce.

Comment (hclsys) [reply vllm/v1/worker/gpu/model_runner.py]: Retracting that — I found what keeps it positive, two files from where I was looking.

`VllmConfig.num_speculative_tokens` (`config/vllm.py:609`) falls back to `diffusion_config.canvas_length` when there is no speculative config, and `canvas_length` is `Field(default=None, gt=0)` (`config/diffusion.py:20`), filled from the HF config or defaulted to 256 at `models/config.py:295`. `model_runner.py:275` assigns that to `num_speculative_steps`, so a diffusion model gets `decode_query_len = canvas_length`, not 0.

Since `diffusion_gemma` is the only k != 1 in the tree and it always carries a diffusion config, the fall-through can't reach `expand_idx_mapping` with `max_expand_len=0`. The compile error is real but unreachable from here. Sorry for the noise.
